symbolic-debuginfo/src/elf.rs

```diff
@@ -881,3 +899,98 @@ impl<'data, 'object> Iterator for ElfSymbolIterator<'data, 'object> {
         })
     }
 }
+
+/// Parsed debug link section.
+#[derive(Debug)]
+pub struct DebugLink<'data> {
+    filename: Cow<'data, CStr>,
+    crc: u32,
+}
+
+impl<'data> DebugLink<'data> {
+    /// Attempts to parse a debug link section from its data.
+    ///
+    /// The expected format for the section is:
+    ///
+    /// - A filename, with any leading directory components removed, followed by a zero byte,
+    /// - zero to three bytes of padding, as needed to reach the next four-byte boundary within the section, and
+    /// - a four-byte CRC checksum, stored in the same endianness used for the executable file itself.
+    /// (from <https://sourceware.org/gdb/current/onlinedocs/gdb/Separate-Debug-Files.html#index-_002egnu_005fdebuglink-sections>)
+    ///
+    /// # Errors
+    ///
+    /// If the section data is malformed, in particular:
+    /// - No NUL byte delimiting the filename from the CRC
+    /// - Not enough space for the CRC checksum
+    pub fn from_data(data: Cow<'data, [u8]>, endianity: Endian) -> Result<Self, DebugLinkError> {
+        // Find the zero byte after the filename.
+        let filename_len_with_nul = match data.iter().position(|&c| c == b'\0') {
+            Some(position) => position + 1,
+            None => return Err(DebugLinkError::MissingNul { data }),
+        };
+
+        // let's be liberal and assume that the padding is correct and all 0s,
+        // and just check that we have enough remaining length for the CRC.
+        if data.len() <= filename_len_with_nul + 4 {
+            return Err(DebugLinkError::MissingCrc {
+                filename_len_with_nul,
+                data,
+            });
+        }
+
+        let crc = match endianity {
+            Endian::Little => {
+                u32::from_le_bytes(data[data.len() - 4..].try_into().expect("wrong size"))
```

Comment:
I understand that seeing `expect()` is not a reassuring sight, especially in code by a new contributor :sweat_smile:.

I'll try to refactor the code because I believe we can remove the expects here.

Reply:
I pushed a new version of the diff that removes all `expect()` calls.